Symbolic parameter / TF compatibility, documentation (#282)
XanaduAI/strawberryfields: doc/introduction/circuits.rst

    circuits/glossary

> In Strawberry Fields, photonic quantum circuits are represented as :class:`~strawberryfields.Program`

Author: This is to get rid of an error message when building the docs (when module A does `from B import X`, Sphinx finds the class X in both A and B and gets confused :/ )

    will need to increase the cutoff dimension.

Symbolic parameters
-------------------

The quantum operations can take both numerical and symbolic parameters. The latter fall into two types:

* **Measured parameters**: Certain quantum programs require that
  operations can be conditioned on measurement results obtained during the execution of the
  program. In this case the parameter value is not known until the measurement is made

>   (or simulated). The latest measurement result of qumode ``i`` is referred to as ``q[i].par``.

Reviewer: Is referred the right word here? 'denoted', or even 'available via ``q[i].par``'?

Author: Available might be the best choice

Reviewer: ```suggestion
  (or simulated). The latest measurement result of qumode ``i`` is available via ``q[i].par``.
```

* **Free parameters**: A *parametrized circuit template* is a program that
  depends on a number of free parameters. These parameters are bound to new fixed
  values each time the program is executed.
  The free parameters are created accessed using the :meth:`Program.params` method.

The symbolic parameters can be combined and transformed using basic algebraic operations and

> common mathematical functions in the ``strawberryfields.parameters.parfuncs`` namespace.

Reviewer: I'm not sure I really like the name `parfuncs`. At the same time, the `strawberryfields.parameters` module is not  user-facing apart from this one import.

I think we should import this to top level, and give it a better name, how about the following?

```python
ops.Sgate(1 - sf.math.sin(q[0].par))
```

This matches nicely with TF 2.0, where a lot of the previous top level functions are now in the `math` module (i.e., `tf.math.log`).

Author: Would `sf.math` be a module, or would `math` be a `SimpleNamespace` object inside `sf`? The latter is how `parfuncs` works now.

Reviewer: I was imagining just a `SimpleNamespace` object. However, it would be nice if documentation and autocomplete could continue to work, maybe it needs `__dir__` defined?

Author: Looks like a `SimpleNamespace` object works nicely, it supports autocompletion and the docs are inherited from the wrapped Sympy funcs.

Author: Okay, it's done.
